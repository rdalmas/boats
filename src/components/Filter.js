import React from "react";

//Create a basic filtering solution for boat length (< 15 OR >= 15) and boat year (< 2010 OR >= 2010).

const Filter = () => (
  <div>Filter</div>
);

export default Filter;