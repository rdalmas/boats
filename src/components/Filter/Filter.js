import React, { useContext, useEffect, useRef, useState, useLayoutEffect } from "react";

import { SearchContext } from "../../pages/Search/searchContext";

const Filter = ({ boats }) => {
  useLayoutEffect(() => {
    filterBoats(boats);
  }, []);
  const [filteredBoats, filterBoats] = useContext(SearchContext);
  const [yearFilter, setYearFilter] = useState(2010);
  const [lengthFilter, setLengthFilter] = useState(15);

  useEffect(() => {
    let result = boats.filter(
      boat => 
      (!lengthFilter || boat.length > lengthFilter) &&
      (!yearFilter || boat.year > yearFilter)
    );
    console.log("result => ",result)
    filterBoats(result);
  }, [yearFilter, lengthFilter]);

  const clearFilter = () => {
    filterBoats(boats);
    filterForm.current.reset();
    setLengthFilter(15);
    setYearFilter(2010);
  }

  const filterForm = useRef(null);
  return (
    <div className="container">
      <div className="row no-gutter">Filters</div>
      <div className="col-sm-12">
        <form onReset={clearFilter} ref={filterForm}>
          <div className="form-group">
            <label htmlFor="lengthFilter">Boat length greater than</label>
            <input type="range" className="form-control-range" name="length" min="0" max="100" id="lengthFilter" value={lengthFilter} onChange={(e) => setLengthFilter(e.target.value)} />
          </div>
          <div className="form-group">
            <label htmlFor="yearFilter">Boat year newer than</label>
            <input type="range" className="form-control-range" min="1950" max="2020" name="year" id="yearFilter" value={yearFilter} onChange={(e) => setYearFilter(e.target.value)} />
          </div>
          <button type="reset">Clear filters</button>
        </form>
      </div>
    </div>
  )
}

export default Filter;