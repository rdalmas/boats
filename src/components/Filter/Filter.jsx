import React, { useContext, useEffect, useRef, useState, useMemo } from "react";

import SearchContext from "../../pages/Search/search.context";
import InputRange from "../../styled-components/InputRange";
import FilterLabel from "../../styled-components/FilterLabel";
import useDebounce from "../../hooks/useDebounce";
import useQueryString from "../../hooks/useQueryString";
import { searchDefaults as def } from "../../constants";

const Filter = ({ boats }) => {
  const [yearQs, setYearQs] = useQueryString("year");
  const [lengthQs, setLengthQs] = useQueryString("length");
  const [,filterBoats] = useContext(SearchContext);
  const [yearFilter, setYearFilter] = useState(yearQs);
  const [lengthFilter, setLengthFilter] = useState(lengthQs);
  const dbLengthFilter = useDebounce(lengthFilter, 300);
  const dbYearFilter = useDebounce(yearFilter, 300);

  const boatsArray = useMemo(() =>
    boats.map(boat => ({
      ...boat,
      price: new Intl.NumberFormat('en-EN', { style: 'currency', currency: 'EUR' }).format(boat.price)
    })),
    [boats]
  );

  useEffect(() => {
    let result = boatsArray.filter(
      boat => 
      (!dbLengthFilter || boat.length > dbLengthFilter) &&
      (!dbYearFilter || boat.year > dbYearFilter)
      );
    filterBoats(result);
    setLengthQs(dbLengthFilter);
    setYearQs(dbYearFilter);
  }, [dbLengthFilter, dbYearFilter, boatsArray, filterBoats, setLengthQs, setYearQs]);

  const clearFilter = () => {
    filterBoats(boatsArray);
    filterForm.current.reset();
    setLengthFilter(def.minLength);
    setYearFilter(def.minYear);
  }

  const filterForm = useRef(null);
  return (
    <div className="container">
      <div className="row">
        <div className="col-md-12">
          <div className="row justify-content-center"><h5>Filters</h5></div>
          <div className="col-md-12">
            <form onReset={clearFilter} ref={filterForm}>
              <div className="form-group">
                <FilterLabel htmlFor="lengthFilter">Boat length greater than</FilterLabel>
                <InputRange id="lengthFilter" min={def.minLength} max={def.maxLength} value={lengthFilter} onChange={setLengthFilter} />
              </div>
              <div className="form-group">
                <FilterLabel htmlFor="yearFilter">Boat year newer than</FilterLabel>
                <InputRange id="yearFilter" min={def.minYear} max={def.maxYear} value={yearFilter} onChange={setYearFilter} />
              </div>
              <div className="row no-gutters justify-content-center">
                <button className="btn btn-secondary btn-md" type="reset">Reset</button>
              </div>
            </form>
          </div>
        </div>
      </div>
    </div>
  );
}

export default Filter;