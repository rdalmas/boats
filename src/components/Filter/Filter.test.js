import React from 'react';
import { shallow } from "enzyme";

import { Filter } from './Filter';
import SearchProvider from "../../pages/Search/search.provider"
import { mockBoats } from "../../testUtils/fixtures/data";

describe("Components - Filter", () => {
  let component;
  beforeEach(() => {
    component = shallow(
      <SearchProvider>
        <Filter boats={mockBoats} />
      </SearchProvider>
    );
  });
  it('should render the Filter correctly', () => {   
    expect(component).toMatchSnapshot();
  });
});