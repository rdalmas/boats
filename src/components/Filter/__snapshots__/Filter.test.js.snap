// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components - Filter should render the Filter correctly 1`] = `
<ContextProvider
  value={
    Array [
      Array [],
      [Function],
    ]
  }
>
  <Filter
    boats={
      Array [
        Object {
          "active": true,
          "cabins": 5,
          "country": "United States",
          "guests": 10,
          "id": 1,
          "imageUrl": "",
          "length": 50,
          "locality": "California",
          "marina": "Buff Marina",
          "name": "Alpha",
          "price": 5000,
          "reviews": Object {
            "score": 5,
            "total": 15,
          },
          "skipper": false,
          "type": "Speed boat",
          "year": 2010,
        },
        Object {
          "active": true,
          "cabins": 5,
          "country": "United States",
          "guests": 10,
          "id": 2,
          "imageUrl": "",
          "length": 40,
          "locality": "California",
          "marina": "Buff Marina",
          "name": "Bravo",
          "price": 5000,
          "reviews": Object {
            "score": 5,
            "total": 15,
          },
          "skipper": false,
          "type": "Speed boat",
          "year": 1951,
        },
        Object {
          "active": true,
          "cabins": 5,
          "country": "United States",
          "guests": 10,
          "id": 1,
          "imageUrl": "",
          "length": 30,
          "locality": "California",
          "marina": "Buff Marina",
          "name": "Charlie",
          "price": 5000,
          "reviews": Object {
            "score": 5,
            "total": 15,
          },
          "skipper": false,
          "type": "Speed boat",
          "year": 2019,
        },
      ]
    }
  />
</ContextProvider>
`;
